### PR TITLE
ROCANA-6608: Support linux network protocol stats

### PR DESCRIPTION
**Pull request by adamlamar**
Adds support for network protocol stats by reading `/proc/net/snmp` and `/proc/net/snmp6`.

### CODE REVIEW

File: sigar_interface.go [hunk 1]
```diff
@@ -94,6 +94,57 @@ type FileSystemUsage struct {
...
+	ICMP ICMPStats
+	UDP  UDPStats
+}
+
+type NetProtoV4Stats struct {
+	IP   IPStats
+	ICMP ICMPStats
+	TCP  TCPStats
+	UDP  UDPStats
+}
```
Comment (matthew-dailey) [inline sigar_interface.go:108]: super nit: should V4 be defined before V6, just for sorting sake?

File: sigar_linux.go [hunk 2]
```diff
@@ -137,6 +137,103 @@ func (self *CpuList) Get() error {
 
 	return err
 }
+func (self *NetProtoV6Stats) Get() error {
+	return readFile(Procd+"/net/snmp6", func(line string) bool {
+		fields := strings.Fields(line)
+		if len(fields) != 2 {
```
Comment (matthew-dailey) [inline sigar_linux.go]: nit: maybe include a comment saying the expected records have key and value, and other lines are ignored?

File: sigar_linux.go [hunk 3]
```diff
@@ -137,6 +137,103 @@ func (self *CpuList) Get() error {
...
+		case "Ip6InAddrErrors":
+			self.IP.InAddrErrors, _ = strtoull(fields[1])
+		case "Ip6OutForwDatagrams":
+			self.IP.ForwDatagrams, _ = strtoull(fields[1])
+		case "Ip6InDelivers":
+			self.IP.InDelivers, _ = strtoull(fields[1])
+		case "Ip6OutRequests":
+			self.IP.OutRequests, _ = strtoull(fields[1])
+		case "Ip6InDiscards":
+			self.IP.InDiscards, _ = strtoull(fields[1])
```
Comment (matthew-dailey) [inline sigar_linux.go]: nit: InDiscards is sorted above OutRequests in the struct definition, so maybe move this block above the OutRequests case

File: sigar_linux.go [hunk 4]
```diff
@@ -137,6 +137,103 @@ func (self *CpuList) Get() error {
...
+		case "Icmp6InMsgs":
+			self.ICMP.InMsgs, _ = strtoull(fields[1])
+		case "Icmp6InErrors":
+			self.ICMP.InErrors, _ = strtoull(fields[1])
+		case "Icmp6OutDestUnreachs":
```
Comment (matthew-dailey) [inline sigar_linux.go]: should this be `Icmp6InDestUnreachs`?  Do you also want to record `Icmp6OutDestUnreachs`?

File: sigar_linux.go [hunk 5]
```diff
@@ -137,6 +137,103 @@ func (self *CpuList) Get() error {
...
+func (self *NetProtoV4Stats) Get() error {
+	return readFile(Procd+"/net/snmp", func(line string) bool {
+		fields := strings.Fields(line)
+		if len(fields) < 2 {
+			return true
+		}
+
+		switch fields[0] {
+		case "Ip:":
+			if fields[1] == "Forwarding" || len(fields) < 20 {
```
Comment (matthew-dailey) [inline sigar_linux.go]: nit: should add comments in here saying that lines alternate between describing the following line, and the data itself.  And then this string being first signals it is a description line

File: sigar_linux.go [hunk 6]
```diff
@@ -137,6 +137,103 @@ func (self *CpuList) Get() error {
...
+				return true
+			}
+			self.IP.InReceives, _ = strtoull(fields[3])
+			self.IP.InAddrErrors, _ = strtoull(fields[5])
+			self.IP.ForwDatagrams, _ = strtoull(fields[6])
+			self.IP.InDelivers, _ = strtoull(fields[9])
+			self.IP.OutRequests, _ = strtoull(fields[10])
+			self.IP.InDiscards, _ = strtoull(fields[11])
```
Comment (matthew-dailey) [inline sigar_linux.go]: nit: InDiscards is sorted above OutRequests in the struct definition, so maybe move this assignment above OutRequests

**Review by matthew-dailey — APPROVED** [review]
LGTM @adamlamar after fixing one typo with `Icmp6InDestUnreachs`.  Other comments are all nits.

Do `snmp` and `snmp6` always return records like that, regardless of OS/kernel version?  I'm wondering if the records in `snmp` come in a different order on, say, Ubuntu vs. CentOS, and you may have to read them 2 lines at a time to get the corresponding indexes for the values.

**Comment by adamlamar** [comment]
Thanks for the 👀  @matthew-dailey! I'm adapting some of the work in #19 and you caught a couple inconsistencies in my changes.

I'm pretty sure the interfaces to `snmp` and `snmp6` are consistent. If they changed, they would break tools like `netstat`. I checked a system running a 4.5 kernel and in `snmp` new attributes are just added to the end. With `snmp6`, they just add new key/value pairs that we would ignore.

I was thinking about refactoring `snmp` to use the column headers instead of the hardcoded offset - since you mentioned a concern there, I'll go ahead.

**Comment by adamlamar** [comment]
@matthew-dailey Feel free to take another look at the map-based solution for `/proc/net/snmp`.

File: sigar_linux.go [hunk 7]
```diff
@@ -137,6 +137,161 @@ func (self *CpuList) Get() error {
...
+
+		// The first field is the protocol name
+		protocol := fields[0]
+
+		// The first line with this protocol name defines the column names
+		positions := protocols[protocol]
+
+		// Parse column headers if needed
+		if positions == nil {
+			positions = make(map[int]string)
```
Comment (matthew-dailey) [inline sigar_linux.go]: Should the mapping be from string to int? (`map[string]int`)?  Then the key is the field name and the value is the position.  That way, in the next section you can do something like this:

```
case "Ip:":
  self.IP.InReceives, _ = strtoull(fields[positions["InReceives"]])
  self.IP.InAddrErrors, _ = strtoull(fields[positions["InAddrErrors"]])
  ...
```

This eliminates the need for the `for position, field` loop as well as the `switch positions[position]` statement.  You can reference each value by name.

**Comment by adamlamar** [comment]
@matthew-dailey That does seem more natural. Its a little more handling, but overall easier to grok.

**Review by matthew-dailey — APPROVED** [review]
This looks really good @adamlamar!

**Pull request merged**